// redux-recompose/lib/injections/baseThunkAction/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

function baseThunkAction(_ref) {
  var type = _ref.type,
      target = _ref.target,
      service = _ref.service,
      _ref$payload = _ref.payload,
      payload = _ref$payload === void 0 ? function () {} : _ref$payload,
      _ref$successSelector = _ref.successSelector,
      successSelector = _ref$successSelector === void 0 ? function (response) {
    return response.data;
  } : _ref$successSelector,
      _ref$failureSelector = _ref.failureSelector,
      failureSelector = _ref$failureSelector === void 0 ? function (response) {
    return response.problem;
  } : _ref$failureSelector;
  var selector = typeof payload === 'function' ? payload : function () {
    return payload;
  };
  return {
    prebehavior: function prebehavior(dispatch) {
      return dispatch({
        type: type,
        target: target
      });
    },
    apiCall: function () {
      var _apiCall = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(getState) {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                return _context.abrupt("return", service(selector(getState())));

              case 1:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      function apiCall(_x) {
        return _apiCall.apply(this, arguments);
      }

      return apiCall;
    }(),
    determination: function determination(response) {
      return response.ok;
    },
    success: function success(dispatch, response) {
      return dispatch({
        type: "".concat(type, "_SUCCESS"),
        target: target,
        payload: successSelector(response)
      });
    },
    failure: function failure(dispatch, response) {
      return dispatch({
        type: "".concat(type, "_FAILURE"),
        target: target,
        payload: failureSelector(response)
      });
    }
  };
}

var _default = baseThunkAction;
exports["default"] = _default;
