// redux-recompose/lib/creators/createModalActions/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

function createModalActions(_ref) {
  var type = _ref.type,
      target = _ref.target,
      contentTarget = _ref.contentTarget;
  return {
    open: function open(modalContent) {
      return {
        type: "".concat(type, "_OPEN"),
        target: target,
        payload: (0, _defineProperty2["default"])({}, contentTarget, modalContent)
      };
    },
    close: function close() {
      var emptyModalContent = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
      return {
        type: "".concat(type, "_CLOSE"),
        target: target,
        payload: (0, _defineProperty2["default"])({}, contentTarget, emptyModalContent)
      };
    }
  };
}

var _default = createModalActions;
exports["default"] = _default;
